src/ORM/DataList.php

```diff
@@ -308,72 +313,118 @@ public function distinct($value)
-     * Return a new DataList instance as a copy of this data list with the sort
-     * order set.
+     * Return a new DataList instance as a copy of this data list with the sort order set
+     *
+     * Raw SQL is not accepted, only actual field names can be passed
      *
-     * @see SS_List::sort()
-     * @see SQLSelect::orderby
+     * @param string|array $args
      * @example $list = $list->sort('Name'); // default ASC sorting
-     * @example $list = $list->sort('Name DESC'); // DESC sorting
+     * @example $list = $list->sort('Name ASC, Age DESC');
      * @example $list = $list->sort('Name', 'ASC');
-     * @example $list = $list->sort(array('Name'=>'ASC', 'Age'=>'DESC'));
-     *
-     * @param string|array Escaped SQL statement. If passed as array, all keys and values are assumed to be escaped.
-     * @return static
+     * @example $list = $list->sort(['Name' => 'ASC', 'Age' => 'DESC']);
+     * @example $list = $list->sort('MyRelation.MyColumn ASC')
+     * @example $list->sort(null); // wipe any existing sort
      */
-    public function sort()
+    public function sort(...$args): static
     {
-        $count = func_num_args();
-
+        $count = count($args);
         if ($count == 0) {
             return $this;
         }
-
         if ($count > 2) {
             throw new InvalidArgumentException('This method takes zero, one or two arguments');
         }
-
         if ($count == 2) {
-            $col = null;
-            $dir = null;
-            list($col, $dir) = func_get_args();
-
-            // Validate direction
-            if (!in_array(strtolower($dir ?? ''), ['desc', 'asc'])) {
-                user_error('Second argument to sort must be either ASC or DESC');
+            list($column, $direction) = $args;
+            $sort = [$column => $direction];
+        } else {
+            $sort = $args[0];
+            if (!is_string($sort) && !is_array($sort) && !is_null($sort)) {
+                throw new InvalidArgumentException('sort() arguments must either be a string, an array, or null');
+            }
+            if (is_null($sort)) {
+                // Set an an empty array here to cause any existing sort on the DataLists to be wiped
+                // later on in this method
+                $sort = [];
+            } elseif (empty($sort)) {
+                throw new InvalidArgumentException('Invalid sort paramater');
+            }
+            // If $sort is string then convert string to array to allow for validation
+            if (is_string($sort)) {
+                $newSort = [];
+                // Making the assumption here there are no commas in column names
+                // Other parts of silverstripe will break if there are commas in column names
+                foreach (explode(',', $sort) as $colDir) {
+                    // Using regex instead of explode(' ') in case column name includes spaces
+                    if (preg_match('/^(.+) ([^"]+)$/i', trim($colDir), $matches)) {
```

Comment:
This will still match against something like `"Column Name"` - it's funtionally equivalent to using `explode` for any string with only one space in it.
If the column name has a space in it, and no direction was supplied, it should _not_ match here so that the default direction can be added in the `else` block.

I think we should explicitly be checking for `asc` and `desc` here instead.
```suggestion
                    if (preg_match('/^(.+) (asc|desc)$/i', trim($colDir), $matches)) {
```

Reply:
It won't match `"Column Name"` - it's pinned with `$` at the end of the regex

It would match `Column Name` - however this will fail on `validateSortDirection()`

Reply:
Apologies, I should have omitted the quote marks which were intended to mean "this is the whole string"

My point is that it shouldn't fail on `validateSortDirection()`, because `Column` on its own won't fail that.
So currently passing `Column` is fine but `Column Name` is not fine, even if they're both valid columns.

Reply:
I don't think think we need to over think for edge cases like this.  An exception will still be thrown and it should be pretty obvious at that point they need to wrap their column in quotes since they'll get a 'Invalid sort direction Name' exception message

Even if we change this to `(asc|desc)`, then it'll still fail on the edge case of someone calling naming column `MySortDir ASC`. There's a workaround available which is to wrap the weird column in double quotes e.g. `->sort('"Column Name" ASC')` 

I think it's reasonable to put the onus on end users to do this here - they'll be well used to do extra steps to get things to actually work.  Other parts of Silverstripe will already be broken e.g. calling $myDataObject->{'Column Name'} to get values (I'm assuming that actually works)

Reply:
I disagree but I'll merge it 'cause it's not worth arguing about. Someone can raise an issue when it causes problems for them.